### ISSUE TITLE
Spread operator inside reduce is causing unnecessary allocations and poor performance, benchmarks say there is 4x difference

### ISSUE THREAD

**Issue by Lesiuk** [rayepps/radash#203]
I was wondering why using `counting` inside my API methods is affecting performance so much and found out the solution.

```javascript
const countingOriginal = <T, TId extends string | number | symbol>(
  list: readonly T[],
  identity: (item: T) => TId
): Record<TId, number> => {
  return list.reduce((acc, item) => {
    const id = identity(item)
    return {
      ...acc,
      [id]: (acc[id] ?? 0) + 1
    }
  }, {} as Record<TId, number>)
}
```

```javascript
const countingFast = <T, TId extends string | number | symbol>(
  list: readonly T[],
  identity: (item: T) => TId
): Record<TId, number> => {
  return list.reduce((acc, item) => {
    const id = identity(item)
    acc[id] = (acc[id] ?? 0) + 1;
    return acc;
  }, {} as Record<TId, number>)
}
```

```
xxx@xxx test % deno bench --allow-all
Check file:///Users/xxx/WebstormProjects/test/main_bench.ts
cpu: Apple M1 Pro
runtime: deno 1.29.1 (aarch64-apple-darwin)

file:///Users/sg0227518/WebstormProjects/test/main_bench.ts
benchmark      time (avg)             (min … max)       p75       p99      p995
------------------------------------------------- -----------------------------
fast       112.71 µs/iter  (92.25 µs … 217.08 µs) 127.71 µs 149.92 µs 155.71 µs
original   709.97 µs/iter (685.62 µs … 863.38 µs)    721 µs 814.33 µs    848 µs

summary
  original
   6.3x slower than fast
```

**Comment by rayepps** [comment]
Thanks for bringing this up @Lesiuk 🍻 a few people have brought this up about different functions and we've been working on fixing them as we go (tracking in issue #78).

I'll fix it in `counting` right now so you don't have an issue with it though 👍

**Comment by rayepps** [comment]
Ok @Lesiuk this has been fixed and is published in `v10.3.1` 🚀
